Feat(eos_designs): Enable evpn_gateway for pathfinder deployment using next-hop-self

ayushmittal-arista opened the pull request:
## Change Summary

Allow configuring EVPN gateway on wan_router using next-hop-self

## Related Issue(s)


## Component(s) name

`arista.avd.eos_designs`

## Proposed changes

- Moved hosts file leaf-wan-use-evpn-on-lan, cv-pathfinder-edge-wan-use-evpn-on-lan under SITE_EVPN in eos_designs_unit_tests 
- Testing the case of client directly connected to a spine.

## How to test

* molecule
* `cv-pathfinder` example updated with a new site

## Checklist

### User Checklist


- [ ] Update cv-pathfinder example Diagrams
- [ ] Update cv-pathfinder example README
- [ ] Update WAN how-to documentation

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has been rebased from devel before I start
- [ ] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [ ] My change requires a change to the documentation and documentation have been updated accordingly.
- [ ] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/intended/configs/site1-wan1.cfg:
```diff
@@ -412,6 +376,7 @@ ip extcommunity-list ECL-EVPN-SOO permit soo 192.168.255.3:101
 !
 ip prefix-list PL-LOOPBACKS-EVPN-OVERLAY
    seq 10 permit 192.168.255.0/24 eq 32
+   seq 20 permit 192.168.42.3/32 eq 32
```
we need to remove this and this will solve the issue we have seen this morning @ayushmittal-arista

gmuloc replied:
let's also add 

```suggestion
ip prefix-list PL-LOOPBACK-VTEP
   seq 10 permit 192.168.42.3/32 eq 32
```

gmuloc replied:
by doing this we can support both gateways side by side

gmuloc reviewed (commented):
TODO in this PR:
* fix comment
* catch HA WAN router with EVPN-GW and raise
* Add limitation that we only support EVPN-GW feature on sites with a single WAN router.
* Add limitation in how-to that we don't support EVPN-GW with WAN HA.
* Add Site 4 to Region 3 showing a single router using EVPN GW

Next PR: (@ayushmittal-arista)
* change the way SOO is computed for multiple WAN routers under the same node group. They should all use the same SOO (of the first router in the group) as is done today for HA.
* replace the limitation above to state that we don't support sites with mix of EVPN-GW and not GW
* Sites with HA should still not have more than 2 routers and they cannot be gateways.
* Site 4 can be expanded to  2 WAN routers (non HA) with EVPN-GW

ayushmittal-arista committed:
Feat(eos_designs): Enable evpn_gateway for pathfinder deployment
- Add check that wan gateway can only be enabled if
    node group contains 1 site. (Should we check based on site)
- Add Site4 with 2 l3leafs and 1 wan router
- Enable Wan Gateway on the wan router.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Doc: Update cv-pathfinder example diagrams

gmuloc committed:
Doc: Realign management IPs for site4

gmuloc committed:
Doc: Fix 2 diagrams

ayushmittal-arista committed:
Merge branch 'devel' into a_4865_12

ayushmittal-arista committed:
run molecule

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Doc: Cleanup old artifacts

gmuloc committed:
Doc: Cleanup old artifacts

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/PATHFINDERS.yml:
```diff
@@ -10,6 +12,7 @@ wan_rr:
     - name: pf1
       id: 1
       mgmt_ip: 192.168.17.10/24
+      evpn_role: none
```
do we still need this?

ayushmittal-arista replied:
Why would we not?

Well technically we only need `overlay_routing_protocol: none` as we generate evpn_overlay config if we don't set either of them to none. 

Setting any one to none is enough. 
There is one place where we check evpn_role != server instead of evon_role == client, which is why it's prefered to set `overlay_routing_protocol: none` . 

But for consistency with other routers in the example, I added both. Do you want me to remove it.

gmuloc replied:
yeah it really comes to the fact that it would be good to change the default basically :) 6.0 is around the corner!

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/SITE1.yml:
```diff
@@ -34,6 +34,7 @@ wan_router:
     uplink_switches: [site1-border1, site1-border2]
     uplink_type: p2p-vrfs # (1)!
     bgp_as: 65000
+    evpn_role: none
```
same for all - do we still need this?
```suggestion
    evpn_role: none
```

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/SITE4.yml:
```diff
@@ -0,0 +1,71 @@
+      cv_pathfinder_region: REGION1
+      cv_pathfinder_site: SITE4
+      # Making this site a transit site
+      cv_pathfinder_transit_mode: region # (2)!
```
```suggestion
      # Making this site a transit site
      cv_pathfinder_transit_mode: region
```

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/cv-pathfinder-edge-wan-use-evpn-on-lan-no-overlay-on-lan.cfg:
```diff
@@ -19,7 +19,7 @@ service routing protocols model multi-agent
 !
 router adaptive-virtual-topology
-   topology role edge
+   topology role edge gateway vxlan
```
hmm this seems wrong with the name of the device (no overlay on lan)

gmuloc replied:
ok looking at the logic and wondering how we could make sure we don't configure this. I guess when `overlay_routing_protocol` is not `ebgp` we shoudl not configure a GW

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/SITE_EVPN.yml:
```diff
@@ -0,0 +1,90 @@
+# enabling EVPN on LAN
+overlay_routing_protocol: ebgp
+
+bgp_as: 65000
```
```suggestion
```

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/SITE_EVPN.yml:
```diff
@@ -0,0 +1,90 @@
+    - name: cv-pathfinder-edge-wan-gateway-spine-uplink
+      cv_pathfinder_region: AVD_Land_West
+      cv_pathfinder_site: Site12
+      # Adding uplink to leaf for EVPN
```
```suggestion
      # Adding uplink to spine for EVPN
```

gmuloc commented on ansible_collections/arista/avd/roles/eos_designs/docs/how-to/wan.md:
```diff
@@ -65,6 +65,7 @@ Please familiarize yourself with the Arista WAN terminology before proceeding:
 - All the WAN routers must have a common path-group with at least one WAN route server to be able to inject the default control-plane match statement in the VRF default WAN policy.
 - For the default VRF, routes received over BGP peering configured under tenants in `network_services` will not be automatically advertised to the WAN (they will be advertised toward the LAN if eBGP is used). To advertise them towards the WAN, they need to be injected in EVPN and this can be achieved by adding a route-map to mark them with the site SOO.
 - Internet exit policies are not supported under WAN port-channel interfaces.
+- EVPN WAN gateway is supported only on sites with single WAN Router.
```
let's move this to the previous section (preview) as it requires to use a key in preview (`wan_use_evpn_node_settings_for_lan`) to be triggered.

A section about this should be added int the how-to as well around the LAN part

gmuloc commented on python-avd/pyavd/_eos_designs/shared_utils/wan.py:
```diff
@@ -648,3 +648,16 @@ def is_wan_vrf(self: SharedUtilsProtocol, vrf: EosDesigns._DynamicKeys.DynamicNe
+        if not gateway:
+            return False
+
+        if self.node_group_config and len(self.node_group_config.nodes) != 1:
```
lets add a negative unit test for this

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/overlay/router_adaptive_virtual_topology.py:
```diff
@@ -44,3 +44,6 @@ def router_adaptive_virtual_topology(self: AvdStructuredConfigOverlayProtocol) -
         self.structured_config.router_adaptive_virtual_topology.site._update(name=wan_site.name, id=wan_site.id)
+
+        if self.shared_utils.evpn_wan_gateway:
+            self.structured_config.router_adaptive_virtual_topology._update(gateway_vxlan=True)
```
```suggestion
            self.structured_config.router_adaptive_virtual_topology.gateway_vxlan = True
```

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/overlay/router_bgp.py:
```diff
@@ -117,7 +117,7 @@ def _set_peer_groups(self: AvdStructuredConfigOverlayProtocol) -> None:
 
             # TODO: AVD 6.0.0 remove the check for WAN routers.
-            if self.shared_utils.overlay_evpn_vxlan is True and (not self.shared_utils.is_wan_router or self.inputs.wan_use_evpn_node_settings_for_lan):
+            if self.shared_utils.overlay_evpn_vxlan is True:
```
I don't think this is the correct condition here - we may need to check for gateway instead

gmuloc replied:
as discussed on call the old condition may need to stay - need to investigate why this does not trigger changes in cv-pathfinder molecule devices without the knob set.

ayushmittal-arista replied:
overlay_routing_protocol is returned as none by the utility for wan routers when knob is disabled.

Probably also why I removed this check, as:
without knob, we would never reach here, 
with knob, this would anyways be always true.

ayushmittal-arista replied:
restored the check for now.

gmuloc replied:
ok lets add a comment and remove the TODO then! Thanks for checking

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/overlay/router_bgp.py:
```diff
@@ -176,7 +176,7 @@ def _set_address_family_ipv4(self: AvdStructuredConfigOverlayProtocol) -> None:
 
         # TODO: no elif
-        elif self.shared_utils.overlay_evpn_vxlan is True:
+        if self.shared_utils.overlay_evpn_vxlan is True:
```
same probably need to check for gateway

gmuloc replied:
otherwise I would expect this to show up even on none gateway

ayushmittal-arista replied:
Again, condition only true for gateway evpn. 
as overlay_routing_protoocl and epvn_role would be none o/w

added the check here for now, awaiting your comment

gmuloc replied:
same thanks for checking - let's not keep extra checks when they are not needed. Thanks!

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/underlay/prefix_lists.py:
```diff
@@ -47,6 +47,12 @@ def prefix_lists(self: AvdStructuredConfigUnderlayProtocol) -> None:
 
         self.structured_config.prefix_lists.append_new(name="PL-LOOPBACKS-EVPN-OVERLAY", sequence_numbers=sequence_numbers)
 
+        # IPv4 - PL-DPS-WAN-OVERLAY
```
lets add a comment about what this prefix list does

gmuloc committed:
Doc: Refactor cv-pathdinder doc

gmuloc committed:
Test: Clean up global BGP AS

gmuloc committed:
Test: Clean up global BGP AS

gmuloc committed:
Refactor: Address PR comment

gmuloc committed:
Test: Add negative unit test

gmuloc committed:
Refactor: Make pre-commit happy again

ayushmittal-arista committed:
address comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/documentation/devices/pf2.md:
```diff
@@ -949,6 +962,14 @@ route-map RM-CONN-2-BGP permit 10
 route-map RM-EVPN-EXPORT-VRF-DEFAULT permit 10
    match extcommunity ECL-EVPN-SOO
+!
+route-map RM-EVPN-SOO-IN deny 10
```
do we need the new route-maps on pathfinder devices? I would think not as it seems we are not using them anywhere?

gmuloc replied:
fixed the logic that was now generating this route-map

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/all.yml:
```diff
@@ -35,3 +35,4 @@ default_node_types:
   - node_type: spine
     match_hostnames:
       - .*-cloud
+      - site.*-spine.*
```
```suggestion
```

gmuloc commented on ansible_collections/arista/avd/examples/cv-pathfinder/intended/configs/pf1.cfg:
```diff
@@ -324,6 +324,14 @@ route-map RM-CONN-2-BGP permit 10
 route-map RM-EVPN-EXPORT-VRF-DEFAULT permit 10
    match extcommunity ECL-EVPN-SOO
 !
+route-map RM-EVPN-SOO-IN deny 10
```
adding an unused route-maps we need to investigate and fix this

gmuloc commented on ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/INVALID_DUAL_ROUTER_WAN_EVPN_GW.yml:
```diff
@@ -0,0 +1,82 @@
+---
+# Testing CV pathfinder edge using new toggle wan_use_evpn_node_settings_for_lan
```
this needs to be adjusted

gmuloc commented on ansible_collections/arista/avd/roles/eos_designs/docs/how-to/wan.md:
```diff
@@ -865,6 +868,37 @@ In the situation where the LAN is EBGP but HA is configured over a direct link,
+
+- the LAN routes are received via EVPN
+- Enabling the gateway requires to configure:
+  - `wan_use_evpn_node_settings_for_lan: True`
```
```suggestion
  - `wan_use_evpn_node_settings_for_lan: true`
```

gmuloc commented on ansible_collections/arista/avd/roles/eos_designs/docs/how-to/wan.md:
```diff
@@ -865,6 +868,37 @@ In the situation where the LAN is EBGP but HA is configured over a direct link,
+  - `overlay_routing_protocol: ebgp` for the WAN router
+  - `evpn_role: client` for the WAN router
+  - an EVPN route server should be defined, e.g. using `evpn_route_servers` settings.
+- When all the conditions are met, the WAN router configuration will add the Gateway parameter and inject the following configuration:
```
```suggestion
- When all the conditions are met, the EVPN Gateway using DPS Interconnect is enabled and the following configuration is added to the WAN router:
```

gmuloc commented on ansible_collections/arista/avd/roles/eos_designs/docs/how-to/wan.md:
````diff
@@ -865,6 +868,37 @@ In the situation where the LAN is EBGP but HA is configured over a direct link,
+
+    ```bash
+    router adaptive-virtual-topology
+       topology role edge gateway vxlan # Notice the gateway VXLAN
````
```suggestion
       topology role edge gateway vxlan # (1)!
```

gmuloc commented on ansible_collections/arista/avd/roles/eos_designs/docs/how-to/wan.md:
````diff
@@ -865,6 +868,37 @@ In the situation where the LAN is EBGP but HA is configured over a direct link,
+          neighbor DC1_RR activate
+          neighbor default next-hop-self received-evpn-routes route-type ip-prefix inter-domain
+    ```
+
````
```suggestion
    ```

1. Notice the gateway VXLAN

```

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Doc: Fix documentation comment

gmuloc committed:
Update ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/all.yml

gmuloc committed:
Refactor: Better logic for SOO route-maps

gmuloc committed:
Refactor: Revert changes post explanation

gmuloc commented:
I am ok with the current state of the PR but would rather have 2 other reviewers than myself

gmuloc committed:
Merge branch 'devel' into a_4865_12

ClausHolbechArista commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/SITE1.yml:
```diff
@@ -36,6 +36,8 @@ wan_router:
     uplink_switches: [site1-border1, site1-border2]
     uplink_type: p2p-vrfs # (1)!
     bgp_as: 65000
+    # Disabling EVPN role. This won't be necessary in AVD 6.0.
+    evpn_role: none
```
It would be better if we could avoid this and scope the preview toggle to only concern the site 4 wan routers.

ClausHolbechArista commented on ansible_collections/arista/avd/examples/cv-pathfinder/group_vars/WAN/cv_pathfinder_settings.yml:
```diff
@@ -4,14 +4,20 @@
 wan_mode: cv-pathfinder
 
+# Enable usage of EVPN settings to configure LAN. This is required to enable EVPN Gateway on Site4.
+wan_use_evpn_node_settings_for_lan: true
```
Move this to site4 yml so we can avoid changing stuff on all the others

ClausHolbechArista commented:
Since we are updating the example to use a preview feature, we should add some note or warning in the example about this.

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Doc: Address PR comments

gmuloc committed:
Merge branch 'devel' into a_4865_12

ClausHolbechArista committed:
Merge branch 'devel' into a_4865_12

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/overlay/router_bgp.py:
```diff
@@ -303,6 +305,10 @@ def _set_address_family_evpn(self: AvdStructuredConfigOverlayProtocol) -> None:
             )
 
+        if self.shared_utils.evpn_wan_gateway:
+            self.structured_config.router_bgp.address_family_evpn.neighbor_default.next_hop_self_received_evpn_routes.enable = True
+            self.structured_config.router_bgp.address_family_evpn.neighbor_default.next_hop_self_received_evpn_routes.inter_domain = True
```
```suggestion
            self.structured_config.router_bgp.address_family_evpn.neighbor_default.next_hop_self_received_evpn_routes._update(enable=True, inter_domain=True)
```

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/underlay/route_maps.py:
```diff
@@ -87,6 +87,12 @@ def route_maps(self: AvdStructuredConfigUnderlayProtocol) -> None:
                 )
 
+            if self.shared_utils.evpn_wan_gateway:
+                sequence_80 = EosCliConfigGen.RouteMapsItem.SequenceNumbersItem(
+                    sequence=80, type="permit", match=EosCliConfigGen.RouteMapsItem.SequenceNumbersItem.Match(["ip address prefix-list PL-DPS-WAN-OVERLAY"])
+                )
+                sequence_numbers.append(sequence_80)
```
```suggestion
                sequence_numbers.append_new(
                    sequence=80, type="permit", match=EosCliConfigGen.RouteMapsItem.SequenceNumbersItem.Match(["ip address prefix-list PL-DPS-WAN-OVERLAY"])
                )
```

ClausHolbechArista reviewed (approved):
Minor code style adjustments, but LGTM.

ayushmittal-arista committed:
Update python-avd/pyavd/_eos_designs/structured_config/overlay/router_bgp.py

Co-authored-by: Claus Holbech <holbech@arista.com>

ayushmittal-arista committed:
Update python-avd/pyavd/_eos_designs/structured_config/underlay/route_maps.py

Co-authored-by: Claus Holbech <holbech@arista.com>

ayushmittal-arista committed:
update molecule

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Merge branch 'devel' into a_4865_12

gmuloc committed:
Merge branch 'devel' into a_4865_12
